sbi/simulators/simutils.py

```diff
@@ -2,82 +2,480 @@
-    """Return simulator input output dimension from prior and observed data. 
+def process_pytorch_prior(prior: Distribution,) -> Tuple[Distribution, int, bool]:
+    """Check validity of pytorch prior, return corrected prior.
+    
+    Args:
+        prior: pytorch distribution prior provided by the user.
     
-    Arguments:
-        prior {Distribution} -- pytorch prior distribution with event and batch shapes
-        observed_data {torch.Tensor} -- Observed data point, x0
+    Raises:
+        ValueError: If prior is defined over a scalar variable.
     
     Returns:
-        dim_input [int] -- input dimension of simulator, i.e., parameter vector dimension, event shape.
-        dim_output [int] -- output dimension of simualtor, i.e., dimension of data or summary stats.
+        prior: Pytorch distribution prior. 
+        parameter_dim: event shape of the prior, number of parameters.
+        prior_returns_numpy: False.
+    """
+
+    # reject scalar priors
+    if prior.sample().ndim == 0:
+        raise ValueError(
+            "Detected scalar prior. Please make sure to pass a pytorch prior with "
+            "batch_shape=torch.Size([1]), or event_shape=torch.Size([1])"
+        )
+
+    assert prior.batch_shape in (
+        torch.Size([1]),
+        torch.Size([]),
+    ), f"The prior must have batch shape torch.Size([]) or torch.Size([1]), but has {prior.batch_shape}."
+
+    check_prior_batch_behavior(prior)
+
+    check_for_batch_reinterpretation_extra_d_uniform(prior)
+
+    parameter_dim = prior.sample().numel()
+
+    return prior, parameter_dim, False
+
+
+def check_for_batch_reinterpretation_extra_d_uniform(prior):
+    """Warn when prior is a batch of scalar Uniforms.
+
+    Most likely the user wants to specify a prior on a multi-dimensional parameter
+    rather than several 1D priors at once.
     """
-    # infer parameter dim by sampling once
-    return prior.sample().numel(), observed_data.numel()
 
+    # Note .batch_shape will always work because and is short-circuiting.
+    if isinstance(prior, Uniform) and prior.batch_shape.numel() > 1:
+        raise ValueError(
+            f"""The specified Uniform prior is a prior on *several scalar parameters*
+            (i.e. a batch), not a prior on a multi-dimensional parameter.
+
+            Please use utils.torchutils.BoxUniform if you'd rather put a prior on a 
+            multi-dimensional parameter.
+            """
+        )
 
-def get_simulator_name(simulator_fun, name=None) -> str:
-    """Get or set name of the simulator. 
+
+def check_for_possibly_batched_observations(true_observation: Tensor):
+    """Raise ValueError if dimensionality of data doesn't match requirements.
     
-    Returns function name if no name is given. 
+    The current requirements are that is 2D with a leading batch dimension of size 1.
+    That is, neither multiple, batched observations are supported, 
+    nor a multidimensional observation e.g., an image.
```

Comment:
No, I think the SBI code as it is now, is **not** able to handle matrices. We will have to decide where to flatten, where reshape to original shape, etc.

Reply:
But then it should fail *elsewhere*

Reply:
Should it really? I thought all these checks are there to inform the user how to use SBI. And with this detailed error message she will know that it is not possible yet. 
If we dont catch this then it will give an uninformative error message about some shape mismatch somewhere deep in nflow or so, no?

Reply:
Yes, it is good to get failures close to where the problem is. Perhaps a warning would be a middle ground. Of course I don't want users to get a cryptic message, but I was more expecting that we would get it ourselves in these early times before official release. Ideally we make a test with parameter dim 2.

Reply:
I think the implementation problem here is that we cannot distinguish between multiple batched observations and a single unbatched matrix observation. Both could have shape (2, 2) (after squeeze).

Reply:
What we could do is to allow for 2D observations if the user provides them with a batch dimension, e.g., in shape (1, 2, 2). If, however, (2, 2) is provided we interpret it as a batch of two (1, 2) observations and throw an error?

Reply:
Yes, that is my understanding of what we agreed upon last time - the leading dim will always be interpreted as a batch dim **except** if the input is a vector `v`. In that case the rule would prescribe 'treat as a batch of scalars' but we instead promote it to a single vector (no batch), i.e. `(1, len(v))`.